invenio_vocabularies/contrib/funders/schema.py

```diff
@@ -8,17 +8,96 @@
 
 """Funders schema."""
 
+from functools import partial
+
 from flask_babelex import lazy_gettext as _
-from marshmallow import Schema, fields, validate
-from marshmallow_utils.fields import SanitizedUnicode
+from marshmallow import Schema, ValidationError, fields, post_load, pre_dump, \
+    pre_load, validate, validates_schema
+from marshmallow_utils.fields import IdentifierSet, SanitizedUnicode
+from marshmallow_utils.schemas import IdentifierSchema
+
+from ...services.schema import BaseVocabularySchema
+from .config import funder_schemes
 
 
 class FunderRelationSchema(Schema):
     """Funder schema."""
 
+    name = SanitizedUnicode(
+        validate=validate.Length(min=1, error=_('Name cannot be blank.'))
+    )
+    id = SanitizedUnicode()
+
+    @validates_schema
+    def validate_funder(self, data, **kwargs):
+        """Validates that either id either name are present."""
+        id_ = data.get("id")
+        name = data.get("name")
+        if id_:
+            data = {"id": id_}
+        elif name:
+            data = {"name": name}
+
+        if not id_ and not name:
+            raise ValidationError(
+                _("An existing id or a free text name must be present"),
+                "affiliations"
+            )
+
+
+class FunderSchema(BaseVocabularySchema):
+    """Service schema for funders."""
+
+    pid = SanitizedUnicode(required=True)
     name = SanitizedUnicode(
         required=True,
         validate=validate.Length(min=1, error=_('Name cannot be blank.'))
     )
-    scheme = SanitizedUnicode()
-    identifier = SanitizedUnicode()
+    country = SanitizedUnicode()
+    identifiers = IdentifierSet(fields.Nested(
+        partial(
+            IdentifierSchema,
+            allowed_schemes=funder_schemes,
+            identifier_required=False
+        )
+    ))
+
+    # TODO: the 3 hooks below could be extracted to a PIDField
+    @pre_load(pass_many=False)
+    def load_pid(self, data, **kwargs):
+        """Loads the pid.
+
+        If it does not exist, it tries to extract it from the identifiers.
+        """
+        if data.get("pid"):
+            return data
+
+        identifiers = data.get("identifiers", [])
+        for identifier in identifiers:
+            # FIXME: make scheme configurable
+            if identifier["scheme"] == "ror":
+                data["pid"] = identifier["identifier"]
+                break
+
+        return data
+
+    @post_load
+    def kepp_pid_on_create(self, data, **kwargs):
+        """Remove the PID if already registered.
+
+        If the pid is not removed it would be injected into the DB JSON column.
+        Since the field is only popped on_create by PIDComponent.
```

Comment:
Not a big fan of relying on the schema to keep the DB JSON "clean". Maybe there's something we can configure/change in the PIDComponent that would help?

Reply:
Yes, we could potentially create a custom PIDComponent and pop it in the update hook but:
- It is still at service level (marshamllow and components both are service level)
- when we talked about this we agreed to do marshmallow 😅 but yeah, I dont like it either, just didnt find another solution.

Reply:
Would it make sense to make `pid` a `dump_only` field maybe and validate separately (as done now kind of)?

Reply:
not really, because you still need it during loading (i.e. for creation). We would need something like `dump_only_on_updates`, which is what we achieve here.